### PR TITLE
Test app factory

### PR DESCRIPTION
**Pull request by JFer11**
In this branch, all tests have been refactored to be compatible with the new initialization method.
Also fixed the old issue of test emails, which were previously sent even if the TESTING flag was True.
Finally, mocks were added to test background tasks.

### CODE REVIEW

File: training/controllers/basic_funcionalities.py [hunk 1]
```diff
@@ -20,6 +20,6 @@ def ver():
 @bp.route('/g')
 def test_g():
     if g.user is None:
-        return "No  hay usuario logueado", 411
+        return "No user logged", 411
```
Comment (joaquinco) [inline training/controllers/basic_funcionalities.py]: The status code here should be 401.

Generally speaking, you'd want an endpoint to get the current user information as you do here, but returning an html or a json depeding if it's a nomral web server or an json api.

File: training/tests/test_email_validation.py [hunk 2]
```diff
@@ -1,27 +1,36 @@
...
+
+        # If you ran tests with de IDE, is probably that app.config was not configured properly, so perhaps
+        # some test will not assert
+
         self.app = app.test_client()
-        db.create_all()
-        db.session.commit()
+        with app.app_context():
+            db.create_all()
+            db.session.commit()
```
Comment (joaquinco) [inline training/tests/test_email_validation.py]: This looks the same as `BasicTests.setUp`, you can create a super class for your test cases and perform the setup there.

Comment (joaquinco) [reply training/tests/test_email_validation.py]: Also, can you rename accordingly all the classes that you've named `BasicTest`?

File: training/tests/test_sign_in_up.py [hunk 3]
```diff
@@ -1,76 +1,85 @@
...
+            # We create two users, with same username
+            self.assertEqual(client.post('/register', data=dict(username='Fernando', email='Fernando1@g.com', password='Fernando', confirm='Fernando', accept_tos=True), follow_redirects=True).status_code, 201)
+            self.assertEqual(client.post('/register', data=dict(username='Fernando', email='Fernando2@g.com', password='Fernando', confirm='Fernando', accept_tos=True), follow_redirects=True).status_code, 420)
 
     # Login Tests
     def test_sign_in_one_user_check_session(self):
         with app.test_client() as client:
             # We create a User
-            self.assertEqual(client.post('/index', data=dict(username='Fernando', email='Fernando@g.com', password='Fernando', confirm='Fernando', accept_tos=True), follow_redirects=True).status_code, 201)
+            self.assertEqual(client.post('/register', data=dict(username='Fernando', email='Fernando@g.com', password='Fernando', confirm='Fernando', accept_tos=True), follow_redirects=True).status_code, 201)
```
Comment (joaquinco) [inline training/tests/test_sign_in_up.py]: You can create a user directly in the database and assume it's there, for example in the setUp method. Also, if you're going to use the same block of code in many places please put it in a function.

**Review by joaquinco — CHANGES_REQUESTED** [review]

File: training/controllers/basic_funcionalities.py (hunk 1 above)
Comment (JFer11) [reply training/controllers/basic_funcionalities.py]: Okey, a json response will be added!

File: training/tests/test_email_validation.py (hunk 2 above)
Comment (JFer11) [reply training/tests/test_email_validation.py]: Understood, I will change it on the next PR.

File: training/tests/test_sign_in_up.py (hunk 3 above)
Comment (JFer11) [reply training/tests/test_sign_in_up.py]: I have already created a function that creates a user and validates his email. I probably forgot to change that when I was doing those tests. I will improve it.

File: training/controllers/basic_funcionalities.py [hunk 4]
```diff
@@ -20,6 +20,10 @@ def ver():
...
     if g.user is None:
-        return "No  hay usuario logueado", 411
+        return "No user logged", 401
     else:
-        return str(g.user), 200
+        user_dict = {}
+        for field in g.user.__dict__:
+            if field != '_sa_instance_state':
+                user_dict[field] = g.user.__dict__[field]
+        return user_dict, 200
```
Comment (joaquinco) [inline training/controllers/basic_funcionalities.py]: Maybe it's just better to go back to the previous version. This is not a good way of serializing the user data, besides the endpoint response format is not consistent on all cases.

File: training/controllers/mail.py [hunk 5]
```diff
@@ -53,12 +53,12 @@ def send_email():
         msg.body = message_body
         if 'send_now' in request.form:
             mail.send(msg)
-            return render_template('email_form_template.html', form=form, sended=True)
+            return render_template('email_form_template.html', form=form, sended=True), 204
         else:
             current_app.task_queue.enqueue('training.background_tasks.tasks.send_email_function', msg)
-            return render_template('email_form_template.html', form=form, sended=True)
+            return render_template('email_form_template.html', form=form, sended=True), 205
```
Comment (joaquinco) [inline training/controllers/mail.py]: Why are you using status code 205?

File: training/tests/super_class.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
+import unittest
+
+from training.app import app
+from training.extensions import db
+
+
+class setUpAndTearDown(unittest.TestCase):
```
Comment (joaquinco) [inline training/tests/super_class.py]: Class names should start with capital letter

Comment (JFer11) [reply training/tests/super_class.py]: Changed!

File: training/controllers/mail.py (hunk 5 above)
Comment (JFer11) [reply training/controllers/mail.py]: Changed to 200! Was a mistake.

File: training/controllers/basic_funcionalities.py (hunk 4 above)
Comment (JFer11) [reply training/controllers/basic_funcionalities.py]: Okey, I will comment it and we will address it in the next PR.

**Pull request merged**